grapher/text/MarkdownTextWrap.tsx

```diff
@@ -0,0 +1,549 @@
+        // to split based on it, and `0` leads to being exactly in between tokens.
+        return { tokenIndex: 0, tokenStartOffset: 0, breakOffset: 0.0001 }
+    }
+    toHTML(key: number | string): JSX.Element {
+        return <React.Fragment key={key}> </React.Fragment>
+    }
+    toSVG(key: number | string): JSX.Element {
+        return <React.Fragment key={key}> </React.Fragment>
+    }
+    toPlaintext(): string {
+        return " "
+    }
+}
+
+export class IRLineBreak implements IRToken {
+    get width(): number {
+        return 0
+    }
+    getBreakpointBefore(): undefined {
+        return undefined
+    }
+    toHTML(key: number | string): JSX.Element {
+        return <br key={key} />
+    }
+    toSVG(): undefined {
+        // We have to deal with this special case in
+        // whatever procedure does text reflow.
+        return undefined
+    }
+    toPlaintext(): string {
+        return "\n"
+    }
+}
+
+export abstract class IRElement implements IRToken {
+    constructor(public children: IRToken[], public fontParams?: IRFontParams) {}
+
+    @imemo get width(): number {
+        return getLineWidth(this.children)
+    }
+
+    getBreakpointBefore(targetWidth: number): IRBreakpoint | undefined {
+        return getBreakpointBefore(this.children, targetWidth)
+    }
+
+    splitBefore(maxWidth: number): {
+        before: IRToken | undefined
+        after: IRToken | undefined
+    } {
+        const { before, after } = splitLineAtBreakpoint(this.children, maxWidth)
+        return {
+            // do not create tokens without children
+            before: before.length ? this.getClone(before) : undefined,
+            after: after.length ? this.getClone(after) : undefined,
+        }
+    }
+
+    splitOnNextLineBreak(): { before?: IRToken; after?: IRToken } {
+        const index = this.children.findIndex(
+            (token) => token instanceof IRLineBreak
+        )
+        if (index >= 0) {
+            return {
+                before:
+                    // do not create an empty element if the first child
+                    // is a newline
+                    index === 0
+                        ? undefined
+                        : this.getClone(this.children.slice(0, index)),
+                after: this.getClone(this.children.slice(index + 1)),
+            }
+        }
+        return { before: this }
+    }
+
+    abstract getClone(children: IRToken[]): IRElement
+    abstract toHTML(key: number | string): JSX.Element
+    abstract toSVG(key: number | string): JSX.Element
+
+    toPlaintext(): string {
+        return lineToPlaintext(this.children)
+    }
+}
+
+export class IRBold extends IRElement {
+    @imemo get width(): number {
+        const plaintext = this.children
+            .map((token) => token.toPlaintext())
+            .join("")
+        return Bounds.forText(plaintext, {
+            ...this.fontParams,
+            fontWeight: 700,
+        }).width
+    }
```

Comment:
Some context:

I think the weird thing with my implementation is that the correct `FontParams` need to be passed into `IRBold` when instantiating it, it doesn't by itself propagate the bold `FontParams`. 

This is because I decided that the element itself (whether text, bold or whatever) should have access to its own width. If we assume that, then the information for its context (font size, weight, etc) needs to be available for the element, and you can't simply instantiate it independently of its context.

So there is the:
- Rendering logic: just wrap the output in `<strong>` and the DOM will take care of propagating the styles to children. This is all that `IRBold` does by itself.
- Text width logic: `IRBold` doesn't propagate this by itself, we need to make sure we propagate `FontParams` down to the `IRText`, as done here: https://github.com/owid/owid-grapher/blob/facc9e60e794405563c7e7308609d4844f21ec14/grapher/text/MarkdownTextWrap.tsx#L391-L396

This is confusing, because the same information lives in two places basically – both outside and inside `IRBold`. I wonder if there is some way to implement it that makes it simpler.